### PR TITLE
Deprecate arg-less calls to subplot_class_factory (and similar factories)

### PR DESCRIPTION
**Pull request by anntzer**
Explicitly passing the base class is not too onerous; end users should
typically not have to handle the arg-less case because we provide these
classes at the toplevel anyways (`Subplot`, `ParasiteAxes`, etc.), and
this refactor should ultimately help with factoring all these factories
together, in particular to make mpl_toolkits Axes picklable (right now
they aren't) without duplicating the somewhat heavy machinery that makes
Subplots picklable (_picklable_subplot_class_constructor, etc.).

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/mpl_toolkits/axes_grid1/parasite_axes.py [hunk 1]
```diff
@@ -175,6 +179,10 @@ def apply_aspect(self, position=None):
 @functools.lru_cache(None)
 def parasite_axes_auxtrans_class_factory(axes_class=None):
     if axes_class is None:
+        cbook.warn_deprecated(
+            "3.3", message="Support for passing None to "
+            "parasite_axes_auxtrans_class_factory is deprecated; explicitly "
+            "pass the default Axes class instead.")
```
Comment (timhoffm) [inline lib/mpl_toolkits/axes_grid1/parasite_axes.py]: Why not be explicit:
`...; explicitly pass ParasiteAxes instead`.

Same for the other warnings.

Comment (anntzer) [reply lib/mpl_toolkits/axes_grid1/parasite_axes.py]: Done.  For the others the default is Axes (well, whatever Axes is in that current module).

File: lib/matplotlib/axes/_subplots.py [hunk 2]
```diff
@@ -186,6 +186,9 @@ def subplot_class_factory(axes_class=None):
     not have to be created for every type of Axes.
     """
     if axes_class is None:
+        cbook.warn_deprecated(
+            "3.3", message="Support for passing None to subplot_class_factory "
```
Comment (QuLogic) [inline lib/matplotlib/axes/_subplots.py]: Same thing about the alignment here as the other PR.

Comment (anntzer) [reply lib/matplotlib/axes/_subplots.py]: at least for deprecations this pattern is common through the codebase, as can be checked with `git grep -A2 'message='`)

**Comment by anntzer** [comment]
rebased, I think that's the last one for now so I'll let either of you do the merge :)

**Pull request merged**